### PR TITLE
Add dcgm

### PR DESCRIPTION
**Pull request by gabrielcocenza**
Added DCGM exporter without an strategy

### CODE REVIEW

File: src/service.py [hunk 1]
```diff
@@ -324,7 +362,81 @@ def remove_resources(self) -> bool:
...
+    def hw_tools() -> Set[HWTool]:
+        """Return hardware tools to watch."""
+        return {HWTool.DCGM}
+
+    def set_config(self) -> bool:
+        """Set the necessary exporter configurations."""
+        # refresh the channel if necessary
+        snap.add(self.exporter_name, channel=self.channel)
+        self.snap_client.set(config={"dcgm-exporter-address": f":{self.port}"})
+        self.snap_client.restart(list(self.snap_client.services.keys()))
```
Comment (chanchiwai-ray) [inline src/service.py:435]: Be careful with this: this might mess up the logic for `_on_cos_agent_relation_`

File: src/service.py [hunk 2]
```diff
@@ -74,6 +90,28 @@ def validate_exporter_configs(self) -> Tuple[bool, str]:
             return False, "Invalid config: 'exporter-log-level'"
         return True, "Exporter config is valid."
 
+
+class RendarableExporter(BaseExporter):
```
Comment (aieri) [inline src/service.py:94]: in the future we will not have to RenderableExporters, could we try to avoid having to rework the existing exporter classes?

File: src/service.py [hunk 3]
```diff
@@ -324,7 +362,81 @@ def remove_resources(self) -> bool:
...
+class SnapExporter(BaseExporter):
+    """A class representing a snap exporter."""
+
+    exporter_name: str
+
+    def __init__(self, channel: str, port: int):
+        """Init."""
+        self.channel = channel
+        self.port = port
+        self.snap_client = snap.SnapCache()[self.exporter_name]
```
Comment (aieri) [inline src/service.py:374]: I don't like how hardcoded the exporter properties are. Different exporters may have different charm config options pertaining to them. Wouldn't it be simpler to just pass them all the charm config options (in a configure method) and let the specific exporter sort out what they need?

File: src/service.py [hunk 4]
```diff
@@ -324,7 +362,81 @@ def remove_resources(self) -> bool:
...
+        """Restart the exporter daemon."""
+        self.snap_client.restart(list(self.snap_client.services.keys()))
+
+    def check_health(self) -> bool:
+        """Check if all services are active."""
+        return all(service.get("active", False) for service in self.snap_client.services.values())
+
+    def set_config(self) -> bool:
+        """Set the necessary exporter configurations."""
+        return True
```
Comment (aieri) [inline src/service.py:417]: I wouldn't use a boolean to represent success; if the function is called `set_config` (and it's weird that it has no arguments), I would expect None as a return type (and an exception if setting the config fails).

**Review by aieri — COMMENTED** [review]
Although this smells a bit like COU class hierarchies, I generally like it. Hardware observer's code feels to me really hard to follow and overly complex, probably due to the complexities of supporting those binary resources and unpackaged exporters. Going forward we will _only_ have snapped exporters though, so we should be able to remove a ton of complexity. After all, a snap doesn't need much to be handled, probably just installing, configuring (via snap set, not config files), starting/stopping the services (without needing systemd templates), and some way to check the services are alive. From this perspective perhaps the dcgm (and smartctl) snaps can provide a much needed reset.

I have added more specific comments directly in the body of the PR.

File: src/service.py [hunk 5]
```diff
@@ -324,7 +362,81 @@ def remove_resources(self) -> bool:
...
+
+        logger.info("Installed %s from channel: %s", self.exporter_name, self.channel)
+        # enable services because some might be disabled by default
+        self.enable_and_start()
+        return True
+
+    def uninstall(self) -> bool:
+        """Uninstall the snap."""
+        snap.remove([self.exporter_name])
+        return self.snap_client.present is False
```
Comment (jneo8) [inline src/service.py:397]: **nitpick (blocking):** 

Please consider follow the current strategy pattern design, all the external binary/package management should use strategy for consistent. It will be similar to [What we have for smart-exporter](https://github.com/canonical/hardware-observer-operator/blob/d420c6364551fe15021b3ac3fdfb28d9ed41b906/src/service.py#L308-L324)

File: src/service.py (hunk 2 above)
Comment (jneo8) [reply src/service.py:94]: I will proposal to follow the idea on [issue #272] if we want to create another different workflow.

The scope of #272 will be too big for this task since it's a big change. So inheritance can provide more stable and consistent exporter service workflow since we have tested this. So I am ok with what we have in the PR now.

File: src/service.py (hunk 3 above)
Comment (jneo8) [reply src/service.py:374]: +1, in this case `ConfigData` is a better option for model.

File: src/service.py [hunk 6]
```diff
@@ -36,28 +37,43 @@ class ExporterError(Exception):
...
-        """Return hardware tools to watch."""
+    def enable_and_start(self) -> None:
+        """Enable and start the exporter services."""
+
+    @abstractmethod
+    def disable_and_stop(self) -> None:
+        """Disable and stop the exporter services."""
+
+    @abstractmethod
+    def set_config(self) -> bool:
```
Comment (jneo8) [inline src/service.py:75]: **nitpick (non-blocking):** 

`set_config` is not precise since we actually render configuration files on hardware-exporter. Maybe we can use `configure`?

File: src/service.py (hunk 4 above)
Comment (jneo8) [reply src/service.py:417]: This is what our interface design right now, which relate to [workflow in the charm](https://github.com/canonical/hardware-observer-operator/blob/d420c6364551fe15021b3ac3fdfb28d9ed41b906/src/charm.py#L229-L237)

I don't think it's nice to change it on this PR.

**Review by jneo8 — CHANGES_REQUESTED** [review]
Nice work, most of the change are LGTM.

I still want to keep to strategy design, yes just a wrapper,  at this point until we finish the #272 .

File: src/service.py (hunk 4 above)
Comment (gabrielcocenza) [reply src/service.py:417]: I agree with Andrea, but as James said is the current implementation and changing here the behavior would increase a lot the changes

File: src/service.py (hunk 1 above)
Comment (gabrielcocenza) [reply src/service.py:435]: I didn't understand...
On `_on_cos_agent_relation_joined` it's going to `enable_and_start` and on `_on_cos_agent_relation_departed` is going to `disable_and_stop`. What exactly can be affected with this?

File: src/service.py (hunk 5 above)
Comment (gabrielcocenza) [reply src/service.py:397]: `install_resources`, `resources_exist` shouldn't be public methods because they are used **only** on  `install`. The snaps does not have resources to be installed per say, so this is more a rendarable characteristic from packages that are not snaps.

The same applies for `remove_resources` that is just used on the `uninstall` that is different on snaps

Would you be ok transforming those methods into "private" of the `RendarableExporter` class to make it clear?

File: src/charm.py [hunk 7]
```diff
@@ -37,6 +37,7 @@ def __init__(self, *args: Any) -> None:
             metrics_endpoints=[
                 {"path": "/metrics", "port": int(self.model.config["hardware-exporter-port"])},
                 {"path": "/metrics", "port": int(self.model.config["smartctl-exporter-port"])},
+                {"path": "/metrics", "port": int(self.model.config["dcgm-exporter-port"])},
```
Comment (gabrielcocenza) [inline src/charm.py:40]: @aieri I'm keeping the port config because how we are handling this to pass to grafana-agent.

**Comment by gabrielcocenza** [comment]
> I still want to keep to strategy design, yes just a wrapper

@jneo8 I don't see a strong reason why to add strategy  snap exporters. They are working fine this way and if add the strategy, we would need to inject the config to `HWToolHelper` to instantiate a strategy. Than we would need to map the install from the strategy to the install from the exporter and so on...

File: src/service.py (hunk 2 above)
Comment (aieri) [reply src/service.py:87]: Now I better understand the flow here and I see how adding RenderableExporter makes sense

File: src/charm.py (hunk 7 above)
Comment (aieri) [reply src/charm.py:40]: remember that it's really hard to remove a config option. I would prefer to pass a hardcoded value for now

File: src/charm.py [hunk 8]
```diff
@@ -82,6 +83,9 @@ def exporters(self) -> List[BaseExporter]:
         if stored_tools & SmartCtlExporter.hw_tools():
             exporters.append(SmartCtlExporter(self.charm_dir, self.model.config))
 
+        if DCGMExporter.hw_tools() & stored_tools:
```
Comment (aieri) [inline src/charm.py:86]: nit: please flip the order here to be consistent with the line above

File: src/service.py [hunk 9]
```diff
@@ -36,6 +37,56 @@ class ExporterError(Exception):
...
+
+    @abstractmethod
+    def disable_and_stop(self) -> None:
+        """Disable and stop the exporter services."""
+
+    @abstractmethod
+    def configure(self) -> bool:
+        """Set exporter config."""
+
+    def validate_exporter_configs(self) -> Tuple[bool, str]:
```
Comment (aieri) [inline src/service.py:78]: I think this should be done at the charm config parsing level, but I understand it's out of scope of this PR

File: src/service.py [hunk 10]
```diff
@@ -36,6 +37,56 @@ class ExporterError(Exception):
...
+        """Validate the static and runtime config options for the exporter."""
+        if not 1 <= self.port <= 65535:
+            logger.error("Invalid exporter port: port must be in [1, 65535].")
+            return False, "Invalid config: exporter's port"
+
+        return True, "Exporter config is valid."
+
+
+class RendarableExporter(BaseExporter):
```
Comment (aieri) [inline src/service.py:87]:
```suggestion
class RenderableExporter(BaseExporter):
```

File: src/service.py [hunk 11]
```diff
@@ -36,6 +37,56 @@ class ExporterError(Exception):
...
+    """A class representing a exporter that needs to render its configuration."""
```
Comment (aieri) [inline src/service.py:88]:
```suggestion
    """A class representing an exporter that needs to render its configuration."""
```

File: src/service.py (hunk 4 above)
Comment (aieri) [reply src/service.py:442]: understood. I've now reviewed more carefully and see the pattern ( :cry: )

**Comment by jneo8** [comment]
> @jneo8 I don't see a strong reason why to add strategy snap exporters. They are working fine this way and if add the strategy, we would need to inject the config to `HWToolHelper` to instantiate a strategy. Than we would need to map the install from the strategy to the install from the exporter and so on...

Please check the source code how [SmartExporter](https://github.com/canonical/hardware-observer-operator/blob/d420c6364551fe15021b3ac3fdfb28d9ed41b906/src/service.py#L266-L324) use the strategy. You don't have to use `HWToolHelper`, you can register strategy on the service class directly and use it. The design idea is the package install/remove logic should be define in the strategy class and service class only handle the service life-cycle.

File: src/service.py [hunk 12]
```diff
@@ -324,7 +370,106 @@ def remove_resources(self) -> bool:
...
+        self.snap_client = snap.SnapCache()[self.exporter_name]
+
+    @staticmethod
+    def hw_tools() -> Set[HWTool]:
+        """Return hardware tools to watch."""
+        return set()
+
+    def install(self) -> bool:
+        """Install the snap from a channel."""
+        if not self._install() or not self.configure():
```
Comment (aieri) [inline src/service.py:391]: non-blocking: having side-effects in an if condition feels awkward to me, but I'll leave this comment as non-blocking as the only alternative I can think of is also rather ugly
```
success = self._install()
if not success:
    return False

success = self.configure()
if not success:
    return False
```
yuck...

File: src/service.py [hunk 13]
```diff
@@ -324,7 +370,106 @@ def remove_resources(self) -> bool:
...
+    @staticmethod
+    def hw_tools() -> Set[HWTool]:
+        """Return hardware tools to watch."""
+        return {HWTool.DCGM}
+
+    def configure(self) -> bool:
+        """Set the necessary exporter configurations."""
+        # refresh the channel if necessary
+        if not self._install():
+            return False
```
Comment (aieri) [inline src/service.py:465]: shouldn't this be part of the `SnapExporter` class?

File: tests/unit/test_service.py [hunk 14]
```diff
@@ -14,7 +15,7 @@
 from config import HARDWARE_EXPORTER_SETTINGS, HWTool
 
 
-class TestBaseExporter(unittest.TestCase):
+class TestRendarableExporter(unittest.TestCase):
```
Comment (aieri) [inline tests/unit/test_service.py:18]:
```suggestion
class TestRenderableExporter(unittest.TestCase):
```

File: src/service.py (hunk 1 above)
Comment (chanchiwai-ray) [reply src/service.py:468]: Sorry, I didn't remember correctly (still remember the old version...), the config_changed are now protected by `if self.cos_agent_related`, so ignore my comment. I remember in the past, the exporters will be restarted even if `cos_agent` does not exist...

File: tests/unit/test_service.py [hunk 15]
```diff
@@ -36,9 +37,9 @@ def setUp(self) -> None:
             "redfish-password": "",
         }
         self.mock_stored_hw_available = {"storcli", "ssacli"}
-        service.BaseExporter.__abstractmethods__ = set()
+        service.RendarableExporter.__abstractmethods__ = set()
```
Comment (aieri) [inline tests/unit/test_service.py:40]:
```suggestion
        service.RenderableExporter.__abstractmethods__ = set()
```

File: tests/unit/test_service.py [hunk 16]
```diff
@@ -36,9 +37,9 @@ def setUp(self) -> None:
...
 
-        self.exporter = service.BaseExporter(
+        self.exporter = service.RendarableExporter(
```
Comment (aieri) [inline tests/unit/test_service.py:42]:
```suggestion
        self.exporter = service.RenderableExporter(
```

File: src/service.py (hunk 5 above)
Comment (jneo8) [reply src/service.py]: Agree, I am ok this the change. I should make my comment more clear, we should use strategy instead of define snap install/remove logic on the service class.

**Review by aieri — CHANGES_REQUESTED** [review]
Thanks, this is looking good but I have a few comments

File: src/service.py (hunk 5 above)
Comment (aieri) [reply src/service.py]: Ah I understand now... Yeah I think for consistency it would make sense, although I think that in the medium term we would _only_ have snap strategies, and at that point the complexity tradeoff of using the strategy pattern (vs a simpler class hierarchy structure) may no longer be worth it

Comment (jneo8) [reply src/service.py]: I still believe the Strategy pattern is valuable here because it allows us to define a generic service lifecycle for hardware, smart, and DCGM exporters. By encapsulating the resource-specific logic (such as handling the installation and removal of deb packages, vendor tools, or snap packages) within the Strategy pattern, we can hide the complexity of these operations. This keeps the core lifecycle management clean and focused on the high-level workflow, while making it easier to extend and maintain when new types of resources or exporters are introduced.

**Comment by gabrielcocenza** [comment]
Still need to manually test to see if it's behaving as expected, but I've included the strategy that @jneo8 asked and another round of review is needed.

 I don't fully agree that this is the right approach, but let's discuss that on a major refactor in another PR.

File: src/service.py [hunk 17]
```diff
@@ -36,6 +37,56 @@ class ExporterError(Exception):
...
+    log_level: str
+
+    @staticmethod
+    @abstractmethod
+    def hw_tools() -> Set[HWTool]:
+        """Return hardware tools to watch."""
+
+    @abstractmethod
+    def install(self) -> bool:
+        """Install the exporter."""
```
Comment (samuelallan72) [inline src/service.py:52]: Please document the meaning of return value in all these methods. :)

File: src/hw_tools.py [hunk 18]
```diff
@@ -615,6 +667,10 @@ def detect_available_tools() -> Set[HWTool]:
 class HWToolHelper:
     """Helper to install vendor's or hardware related tools."""
 
+    def __init__(self, config: ConfigData) -> None:
```
Comment (jneo8) [inline src/hw_tools.py:670]: **nitpick (blocking):** 

This is not needed. Please take a look the different between the `SmartCtlExporter` and `SmartCtlExporterStrategy`, how hw-observer control their life-cycle.

- [Where we register SmartCtlStrategy](https://github.com/canonical/hardware-observer-operator/blob/1396dbd97139d55ed44d10a696c732efcf602d84/src/hw_tools.py#L631)

- [Where we register SmartCtlExporterStrategy](https://github.com/canonical/hardware-observer-operator/blob/1396dbd97139d55ed44d10a696c732efcf602d84/src/service.py#L278C25-L278C49)

**Review by jneo8 — CHANGES_REQUESTED** [review]

File: src/hw_tools.py (hunk 18 above)
Comment (jneo8) [inline src/hw_tools.py:670]: **nitpick (blocking):** 

The change on HWToolHelper is not needed. Please take a look the different between the `SmartCtlExporter` and `SmartCtlExporterStrategy`, how hw-observer control their life-cycle.

- [Where we register SmartCtlStrategy](https://github.com/canonical/hardware-observer-operator/blob/1396dbd97139d55ed44d10a696c732efcf602d84/src/hw_tools.py#L631)

- [Where we register SmartCtlExporterStrategy](https://github.com/canonical/hardware-observer-operator/blob/1396dbd97139d55ed44d10a696c732efcf602d84/src/service.py#L278C25-L278C49)

**Review by jneo8 — CHANGES_REQUESTED** [review]

File: src/hw_tools.py (hunk 18 above)
Comment (gabrielcocenza) [reply src/hw_tools.py:670]: Can you be more specific why it's not needed?

SmartCtl is being treated as a debian package and there is no charm configuration needed during the installation. With a snap I need to pass the configuration to know what is the channel to install from. 
How am I suppose to install the snap in the strategy without passing the channel? Creating a dummy install on the strategy and install for real in the exporter itself?

File: src/service.py [hunk 19]
```diff
@@ -324,7 +370,98 @@ def remove_resources(self) -> bool:
...
+    def restart(self) -> None:
+        """Restart the exporter daemon."""
+        self.snap_client.restart(reload=True)
+
+    def check_health(self) -> bool:
+        """Check if all services are active.
+
+        Returns true if the service is healthy, false otherwise.
+        """
+        return self.strategy.check()
```
Comment (chanchiwai-ray) [inline src/service.py:437]: It's weird to see that `self.strategy.check` is used here. The `check` in strategy class should be checking if the installation is successful of not; while the `check` here in the exporter should be checking the exporter's health.

File: src/hw_tools.py [hunk 20]
```diff
@@ -187,6 +188,57 @@ def remove(self) -> None:
...
+        except Exception as err:  # pylint: disable=broad-except
+            logger.error("Failed to remove %s: %s", self.snap, err)
+            raise err
+
+    def check(self) -> bool:
+        """Check if all services are active."""
+        return all(
+            service.get("active", False)
+            for service in snap.SnapCache()[self.snap].services.values()
+        )
```
Comment (chanchiwai-ray) [inline src/hw_tools.py:229]: Based on other strategy implementation, shouldn't this checking the installation status?

File: src/service.py (hunk 19 above)
Comment (jneo8) [reply src/service.py:437]: Maybe both?

check resource health then service health.

**Review by jneo8 — APPROVED** [review]
Thanks, after address those comments, it's LGTM.

**Review by aieri — APPROVED** [review]

File: src/service.py (hunk 19 above)
Comment (gabrielcocenza) [reply src/service.py:437]: > in strategy class should be checking if the installation is successful or not;

Isn't the responsibility of the install method itself to check if the installation was successful. I mean what is necessary to check if the installation doesn't throw an exception and when you ask to check the services are healthy? (meaning that is installed, otherwise you couldn't check) 

I think this is going to change in a near future when the charm will install the NVIDIA drivers, but this change will be placed at the DCGM class and not on the Snap class itself. In any case, I don't have a strong opinion and I would be ok changing the behavior.

As you didn't block the PR, I'm solving this conversation, but feel free to open a new PR changing this behavior if necessary

File: src/hw_tools.py (hunk 20 above)
Comment (gabrielcocenza) [reply src/hw_tools.py:229]: Solving as explained in the comment above, but feel free to change it if necessary.
